### PR TITLE
GitHub Actions Sanitizers Checks: ubsan

### PR DESCRIPTION
**Pull request by williamfgc**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Enable sanitizer checks using clang compilers as a GitHub Actions matrix for: asan, ubsan, tsan and msan for deterministic tests. The goal is to catch more memory-related bugs and guide fixes.

## What type(s) of changes does this code introduce?

- Testing changes (e.g. new unit/integration/performance tests)  sanitizers checks

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Local Ubuntu 20.04, the goal is to test it on GitHub Actions runners

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted (NA)
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

Related to #3020

### CODE REVIEW

**Comment by prckent** [comment]
William - I see that these are running so the answer to the following question will soon be clear, but I was wondering if you have been able to do a test run, and if so, what do the results look like in your opinion? i.e. Do you think that they are actionable as is, need to be redirected to another service, need some whitelisting to be useful/useable?

File: CMakeLists.txt [hunk 1]
```diff
@@ -267,6 +267,23 @@ CMAKE_DEPENDENT_OPTION(USE_VTUNE_TASKS "USE VTune ittnotify task annotation" OFF
 # CMake note - complex conditionals in cmake_dependent_option must have spaces around parentheses
 CMAKE_DEPENDENT_OPTION(USE_NVTX_API "Enable/disable NVTX regions in CUDA code." OFF "ENABLE_TIMERS AND ( QMC_CUDA OR ENABLE_CUDA )" OFF)
 
+######################################################################
+# Sanitizer options
+######################################################################
+
+# Add optional sanitizers ASAN, UBSAN, MSAN
+SET(ENABLE_SANITIZER "NONE" CACHE STRING "NONE,ASAN,UBSAN,MSAN")
+SET_PROPERTY(CACHE ENABLE_SANITIZER PROPERTY STRINGS "NONE" "ASAN" "UBSAN" "MSAN")
```
Comment (ye-luo) [inline CMakeLists.txt:276]: Is it better to have "none" "asan" "ubsan" "msan" all in lower case?

File: CMakeLists.txt [hunk 2]
```diff
@@ -267,6 +267,23 @@ CMAKE_DEPENDENT_OPTION(USE_VTUNE_TASKS "USE VTune ittnotify task annotation" OFF
...
+
+# Perform sanitizer option check, only works in debug mode
+IF( NOT "${ENABLE_SANITIZER}" STREQUAL "NONE")
+  MESSAGE( STATUS "Enable sanitizer ENABLE_SANITIZER=${ENABLE_SANITIZER}" )
+  IF( "${ENABLE_SANITIZER}" STREQUAL "ASAN" AND NOT "${CMAKE_BUILD_TYPE}" STREQUAL "DEBUG" )
+    # better to fail than override debug mode
+    MESSAGE( FATAL_ERROR "-DENABLE_SANITIZER=ASAN can only be enabled with -DCMAKE_BUILD_TYPE=Debug" )
```
Comment (ye-luo) [inline CMakeLists.txt:283]: To my understanding, this is not required. Release, Debug, RelWithDebugInfo all work but the level of info is different. RelWithDebugInfo is my recomendataion not Debug.

File: CMakeLists.txt (hunk 1 above)
Comment (ye-luo) [reply CMakeLists.txt:276]: If I give a random string ("abc") not listed above, does cmake stop?

File: CMakeLists.txt [hunk 3]
```diff
@@ -865,6 +882,30 @@ IF(ENABLE_GCOV)
...
+        CACHE STRING "UndefinedBehaviorSanitizer C++ compiler builds." FORCE)
+  ELSEIF( "${ENABLE_SANITIZER}" STREQUAL "MSAN" )
+    SET(CMAKE_CXX_FLAGS_SAN "-fsanitize=memory"
+        CACHE STRING "MemorySanitizer C++ compiler builds." FORCE)
+  ELSEIF( "${ENABLE_SANITIZER}" STREQUAL "TSAN" )
+    SET(CMAKE_CXX_FLAGS_SAN "-fsanitize=thread"
+        CACHE STRING "ThreadSanitizer C++ compiler builds." FORCE)
+  ENDIF()
+  
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CMAKE_CXX_FLAGS_SAN}")
```
Comment (ye-luo) [inline CMakeLists.txt:905]: Please protect the case of none namely CMAKE_CXX_FLAGS_SAN not set to avoid adding a white space.

**Review by ye-luo — COMMENTED** [review]
Need to clean up previous sanitizer options in CMake/ClangCompilers.cmake
and update manual entry.

**Comment by williamfgc** [comment]
@prckent good points, I'd merge `ubsan` as-is (good job on that!) as it's low-overhead. `tsan` seems to be just hanging on GitHub Actions. `msan` is catching some stuff from Catch (pun not intended), so I need to set a filter of some sort.The most useful is `asan` as I see something that looks like a "true positive" from using `new` in `/src/Estimators/LocalEnergyEstimator.cpp:38`.

My preference would be to keep all the CMake configuration, and:
- Merge `ubsan`  as-is
- Deactivate `asan` (priority to resolve), `msan`, `tsan` on GitHub Actions until issues are resolved.

The above is to keep CI clean and short. Let me know what you think, then we can merge and start chasing leaks.

File: CMakeLists.txt (hunk 1 above)
Comment (williamfgc) [reply CMakeLists.txt:276]: @ye-luo I can change it to lower case if that's the preference. Good catch on the second issue, I can make it error out if the entry is not valid. `SET_PROPERTY` is for `ccmake` or `cmake-gui`.

File: CMakeLists.txt (hunk 2 above)
Comment (williamfgc) [reply CMakeLists.txt:283]: Will do.

**Comment by prckent** [comment]
@williamfgc I agree with your recommendations. We put ubsan into production immediately , then work to fix the code for asan to pass and get that in production. asan actually looked quite close. msan is where the big work is imo, but I suspect there are many instances of the same error being repeated over and over.

After a few more PRs have run through I think we can set ubsan and the original linux build action as required to pass.

File: CMakeLists.txt (hunk 3 above)
Comment (williamfgc) [reply CMakeLists.txt:904]: Good line, this line should be protected as it's executed inside a `if not equal none` block, but please double-check I'm not missing anything else. Thanks!

**Comment by prckent** [comment]
OK to test

**Review by prckent — APPROVED** [review]
LGTM

Ye to merge or request further changes

**Review by ye-luo — APPROVED** [review]
LGTM

**Comment by ye-luo** [comment]
@williamfgc could you relocate config/sanitizers/lsan.supp under tests/test_automation/github-actions/ci?

**Comment by williamfgc** [comment]
@ye-luo just double-checking 'cause the file is not exclusive to github-actions/ci, is this still preferred?

**Comment by prckent** [comment]
If my understanding is correct, I think it is not plumbed in to regularly sanitizer builds yet, but might be in future. In which case leave where it is. (We would need to configure ASAN_OPTIONS=suppression=${GITHUB_WORKSPACE}/config/sanitizers/lsan.supp ?)

**Comment by williamfgc** [comment]
@prckent yes you're correct, it's not currently part of the GitHub Actions runners until leaks are resolved. It's needed locally when debugging leaks to avoid false positives. Yes, `LSAN_OPTIONS` has to be configured to this file (the same with `ASAN_OPTIONS` if we need one), see [here](https://github.com/QMCPACK/qmcpack/blob/develop/tests/test_automation/github-actions/ci/run_step.sh#L80). Thanks!

**Comment by ye-luo** [comment]
As long as it is primarily used/maintained for github-actions/ci, it is better to keep this file with other files for the same purpose. I'm trying to keep `/config` folder for build recipes instead of any kind of configuration files. No need to move the file right now, just keep in mind and do it when other changes are needed for github actions.

**Pull request merged**